test/integration/ddl_sharding_key_test.lua

```diff
@@ -0,0 +1,181 @@
+local fio = require('fio')
+
+local t = require('luatest')
+
+local helpers = require('test.helper')
+
+local ok = pcall(require, 'ddl')
+if not ok then
+    t.skip('Lua module ddl is required to run test')
+end
+
+local pgroup = t.group('ddl_sharding_key', {
+    {engine = 'memtx'},
+    {engine = 'vinyl'},
+})
+
+pgroup.before_all(function(g)
+    g.cluster = helpers.Cluster:new({
+        datadir = fio.tempdir(),
+        server_command = helpers.entrypoint('srv_ddl'),
+        use_vshard = true,
+        replicasets = helpers.get_test_replicasets(),
+        env = {
+            ['ENGINE'] = g.params.engine,
+        },
+    })
+    g.cluster:start()
+    local result, err = g.cluster.main_server.net_box:eval([[
+        local ddl = require('ddl')
+
+        local ok, err = ddl.get_schema()
+        return ok, err
+    ]])
+    t.assert_equals(type(result), 'table')
+    t.assert_equals(err, nil)
+end)
+
+pgroup.after_all(function(g) helpers.stop_cluster(g.cluster) end)
+
+pgroup.before_each(function(g)
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key_uniq_index')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_name_key_non_uniq_index')
+    helpers.truncate_space_on_cluster(g.cluster, 'customers_secondary_idx_name_key')
+end)
+
+pgroup.test_select = function(g)
+    local tuple = {2, box.NULL, 'Ivan', 20}
+
+    -- insert tuple
+    local result, err = g.cluster.main_server.net_box:call('crud.insert', {
+        'customers_name_key', tuple
+    })
+
+    t.assert_equals(err, nil)
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+
+    local conditions = {{'==', 'id', 2}}
+
+    local result, err = g.cluster.main_server.net_box:call('crud.select', {
+        'customers_name_key', conditions,
+    })
+
+    t.assert_equals(err, nil)
+    t.assert_not_equals(result, nil)
+    t.assert_equals(#result.rows, 1)
+end
```

Comment:
It seems, the test would work even if crud would use a primary key as sharding key: here we don't verify that correct sharding key is used for insert and select.

I understand it so: you want to verify correctness of select assuming correctness of insert. Right?

If it is so, please, leave a comment: it was not obvious for me.

For verifying select I would prefer an insert operation that does not use crud: use vshard.router.call() directly. This way the test will check select more naturally and will not pass successfully if select and insert will occasionally become broken in the same way (say, due to a mistake in sharding_key.lua).

I propose to do one of the following:

1. Add an explanation what the test case assume and what intend to verify.
2. Rewrite inserting without crud.

I would prefer 2, but I don't insist.

Reply:
(It is from commit 'Support custom sharding key in select'.)

Reply:
I would agree that it was bad idea to test CRUD using CRUD, which you described in comment above.
I have rewrote most part of tests and now I'm using remote box operations for preparing storages in replicaset for testing and use remote box operations for verifying tuples after an action.

Reply:
Moreover, you propose to use `vshard.call()` for setup state before test, but I'm using box operations, and it is even better: we will have more confidence that combo of crud and vshard works properly.

Reply:
> Add an explanation what the test case assume and what intend to verify.

I believe it is a clear testcase description.

```
+-- The main purpose of testcase is to verify that CRUD will calculate bucket_id
+-- using secondary sharding key (name) correctly and will get tuple on storage
+-- in replicaset s2.
+-- bucket_id was calculated using function below:
+--     function(key)
+--         return require('vshard.hash').strcrc32(key) % 3000 + 1
+--     end
+-- where 3000 is a default number of buckets used in vshard.
```

Reply:
Nice!